specfile/conditions.py

```diff
@@ -0,0 +1,141 @@
+            )
+            # parse only once
+            expand.skip_parsing = True
+            return result
+        return Macros.expand(s)
+
+    if keyword in ("%if", "%elif"):
+        try:
+            result = expand(f"%{{expr:{expression}}}")
+        except RPMException:
+            return False
+        try:
+            return int(result) != 0
+        except ValueError:
+            return True
+    elif keyword.endswith("arch"):
+        target_cpu = expand("%{_target_cpu}")
+        match = any(t for t in expression.split() if t == target_cpu)
+        return not match if keyword == "%ifnarch" else match
+    elif keyword.endswith("os"):
+        target_os = expand("%{_target_os}")
+        match = any(t for t in expression.split() if t == target_os)
+        return not match if keyword == "%ifnos" else match
+    return False
+
+
+def process_conditions(
+    lines: List[str],
+    macro_definitions: Optional["MacroDefinitions"] = None,
+    context: Optional["Specfile"] = None,
+) -> List[Tuple[str, bool]]:
+    """
+    Processes conditions in a spec file. Takes a list of lines and returns the same
+    list of lines extended with information about their validity. A line is considered
+    valid if it doesn't appear in a false branch of any condition.
+
+    Args:
+        lines: List of lines in a spec file.
+        macro_definitions: Parsed macro definitions to be used to prevent parsing conditions
+          inside their bodies (and most likely failing).
+        context: `Specfile` instance that defines the context for macro expansions.
+
+    Returns:
+        List of tuples in the form of (line, validity).
+    """
+    excluded_lines = []
+    for md in macro_definitions or []:
+        position = md.get_position(macro_definitions)
+        excluded_lines.append(range(position, position + len(md.body.splitlines())))
+    condition_regex = re.compile(
+        r"""
+        ^
+        \s*                                           # optional preceding whitespace
+        (?P<kwd>%((el)?if(n?(arch|os))?|endif|else))  # keyword
+        \s*
+        (
+            \s+
+            (?P<expr>.*?)                             # expression
+            (?P<end>\s*|\\)                           # optional following whitespace
+                                                      # or a backslash indicating
+                                                      # that the expression continues
+                                                      # on the next line
+        )?
+        $
+        """,
+        re.VERBOSE,
+    )
+    result = []
+    branches = [True]
+    indexed_lines = list(enumerate(lines))
+    while indexed_lines:
+        index, line = indexed_lines.pop(0)
+        # ignore conditions inside macro definition body
+        if any(index in r for r in excluded_lines):
+            result.append((line, branches[-1]))
+            continue
+        m = condition_regex.match(line)
+        if m:
```

Comment:
Maybe a time for walrus operator?..;-D

Reply:
I would also try to use the same trick as above (=reverse the condition and return early) to avoid ging one level deeper to the right.

Reply:
> Maybe a time for walrus operator?..;-D

Not supported in Python 3.6.